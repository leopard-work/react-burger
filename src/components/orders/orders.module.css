.page {
    height: 924px;
}
.wrapper {
    overflow-y: scroll;
    scroll-behavior: smooth;
    height: 100%;
}
.wrapper::-webkit-scrollbar {
    width: 8px;
}
.wrapper::-webkit-scrollbar-track {
    background: #2F2F37;
}
.wrapper::-webkit-scrollbar-thumb {
    background-color: #8585AD;
}
.item {
    border-radius: 40px;
    background: #1C1C21;
    cursor: pointer;
}
.top {
    display: flex;
    flex-direction: row;
    flex-wrap: wrap;
    justify-content: space-between;
    align-items: center;
    align-content: center;
}
.bottom {
    display: flex;
    flex-direction: row;
    flex-wrap: wrap;
    justify-content: space-between;
    align-items: center;
    align-content: center;
}
.params {
    display: flex;
    flex-direction: row-reverse;
    flex-wrap: wrap;
    justify-content: flex-start;
    align-items: center;
    align-content: center;
}
.param {
    width: 64px;
    height: 64px;
    border: solid 2px greenyellow;
    border-radius: 50%;
    display: flex;
    flex-direction: row;
    flex-wrap: wrap;
    justify-content: center;
    align-items: center;
    align-content: center;
    margin-left: -15px;
    background-color: #1c1c21;
}
.param:last-child {
    margin-left: 0px;
}
.price {
    display: flex;
    flex-direction: row;
    flex-wrap: wrap;
    justify-content: flex-start;
    align-items: center;
    align-content: center;
}

