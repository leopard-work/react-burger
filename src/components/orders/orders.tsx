import React, { FC, useEffect } from "react";
import "@ya.praktikum/react-developer-burger-ui-components";
import styles from "./orders.module.css";
import stylesItem from "./orders_item.module.css";
import { CurrencyIcon } from "@ya.praktikum/react-developer-burger-ui-components";
import { useHistory, useLocation, useParams } from "react-router-dom";
import { useAppSelector, useAppDispatch } from "../../services/reducers";
import { ItemProps, OrderItemProps } from "../../utils/types";
import Modal from "../modal/modal";
import {
  CLOSE_ORDER_ITEM,
  OPEN_ORDER_ITEM,
} from "../../services/actions/orders";

export const Orders = (props: { page?: string }) => {
  return (
    <div className={styles.page + " pt-10"}>
      <div className={styles.wrapper + " pr-2"}>
        <OrdersItems page={props.page} />
      </div>
    </div>
  );
};

export const OrdersItems = (props: { page?: string }) => {
  const history = useHistory();
  const location = useLocation();
  const dispatch = useAppDispatch();
  const orders = useAppSelector((state) => state.orders);
  const params: { id: string } = useParams();
  const items = orders.ordersData.orders;
  let page = "feed";
  if (props.page) page = props.page;

  useEffect(() => {
    if (location.state)
      dispatch({
        type: OPEN_ORDER_ITEM,
        item: items.find((item: OrderItemProps) => item._id === params.id),
      });
  }, [location.state, params.id, dispatch, items]);

  const modalClose = () => {
    dispatch({ type: CLOSE_ORDER_ITEM });
    history.push(`/${page}`);
  };

  return (
    <>
      {items.map((item: OrderItemProps) => (
        <OrderItem item={item} page={props.page} key={item._id} />
      ))}
      <Modal isOpen={orders.viewFullOrderModalOpen} close={() => modalClose()}>
        <OrderModalItem item={orders.viewFullOrder} />
      </Modal>
    </>
  );
};

const dateParse = (date: string) => {
  const n = new Date(date);
  return n.toLocaleString();
};

const ingredientsParse = (items: Array<string>, catalog: Array<ItemProps>) => {
  let parseItems: Array<ItemProps> = [];
  let parseItemsElements: Array<JSX.Element> = [];
  let parseItemsElementsModal: Array<JSX.Element> = [];
  let price = 0;
  interface StringArray {
    [index: string]: number;
  }
  let counts: StringArray = {};
  items.map((item) => {
    // @ts-ignore
    parseItems.push(catalog.find((el) => el._id === item));
    return false;
  });
  parseItems.map((item, i: number) => {
    price += item.price;
    if (i < 6) {
      if (i !== 5) {
        parseItemsElements.push(
          <div className={styles.param} key={i}>
            <img src={item.image} alt="" />
          </div>
        );
      } else {
        parseItemsElements.push(
          <div
            className={
              styles.param + " text text_type_main-default text_color_inactive"
            }
            key={i}
          >
            <img src={item.image} alt="" />
            {parseItems.length - 6 !== 0 ? (
              <span className="text text_type_main-default">
                +{parseItems.length - 6}
              </span>
            ) : (
              ""
            )}
          </div>
        );
      }
    }
    counts[item._id] = (counts[item._id] || 0) + 1;
    return false;
  });

  parseItemsElements.reverse();

  const itemsUnique = parseItems.filter(
    (el, ind) => ind === parseItems.indexOf(el)
  );
  itemsUnique.map((item, i) => {
    parseItemsElementsModal.push(
      <div className={stylesItem.param + " mb-4"} key={i}>
        <div className={stylesItem.title}>
          <div className={stylesItem.icon + " mr-4"}>
            <img src={item.image} alt={item.name} />
          </div>
          <p className="text text_type_main-default">{item.name}</p>
        </div>
        <div className={styles.price}>
          <p className="text text_type_digits-default mr-2">
            {counts[item._id]} x {item.price}
          </p>{" "}
          <CurrencyIcon type="primary" />
        </div>
      </div>
    );
    return false;
  });

  return {
    items: parseItemsElements,
    price: price,
    itemsModal: parseItemsElementsModal,
  };
};

const statusParse = (status: string) => {
  if (status === "created") return <span>Создан</span>;
  if (status === "pending") return <span>Готовится</span>;
  if (status === "done") return <span className={styles.done}>Выполнен</span>;
};

const OrderItem = (props: { item: OrderItemProps; page?: string }) => {
  const dispatch = useAppDispatch();
  const history = useHistory();
  let page = "feed";
  if (props.page) page = props.page;
  const openOrder = (order: OrderItemProps) => {
    dispatch({ type: OPEN_ORDER_ITEM, item: order });
    history.replace(`/${page}/${order._id}`, { modal: true });
  };
  const item = props.item;
  const catalog = useAppSelector((state) => state.catalog);
  const parseElements = ingredientsParse(item.ingredients, catalog.items.data);

  return (
    <div
      className={styles.item + " pt-6 pr-6 pb-6 pl-6 mb-4"}
      onClick={() => openOrder(item)}
    >
      <div className={styles.top + " mb-6"}>
        <p className="text text_type_digits-default">#{item.number}</p>
        <p className="text text_type_main-default text_color_inactive">
          {dateParse(item.createdAt)}
        </p>
      </div>
      <p className="text text_type_main-medium mb-2">{item.name}</p>
      <p className="text text_type_main-default mb-6">
        {statusParse(item.status)}
      </p>
      <div className={styles.bottom}>
        <div className={styles.params}>{parseElements.items}</div>
        <div className={styles.price}>
          <p className="text text_type_digits-default mr-2">
            {parseElements.price}
          </p>{" "}
          <CurrencyIcon type="primary" />
        </div>
      </div>
    </div>
  );
};

export const OrderModalItem: FC<{ item: OrderItemProps }> = ({ item }) => {
  const catalog = useAppSelector((state) => state.catalog);
  const parseElements = ingredientsParse(item.ingredients, catalog.items.data);

  return (
    <div className={stylesItem.wrapper + " mt-10 mb-10"}>
      <p className="text text_type_digits-default mb-10">#{item.number}</p>
      <p className="text text_type_main-medium mb-3">{item.name}</p>
      <p className="text text_type_main-default mb-15">
        {statusParse(item.status)}
      </p>
      <p className="text text_type_main-medium mb-6">Состав:</p>
      <div className={stylesItem.params + " pr-6 mb-10"}>
        {parseElements.itemsModal}
      </div>
      <div className={stylesItem.bottom}>
        <p className="text text_type_main-default text_color_inactive">
          {dateParse(item.createdAt)}
        </p>
        <div className={styles.price}>
          <p className="text text_type_digits-default mr-2">
            {parseElements.price}
          </p>{" "}
          <CurrencyIcon type="primary" />
        </div>
      </div>
    </div>
  );
};
