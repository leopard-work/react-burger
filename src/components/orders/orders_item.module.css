.wrapper {
    width: 640px;
    margin-left: auto;
    margin-right: auto;
}
.params {
    overflow-y: scroll;
    scroll-behavior: smooth;
    max-height: 312px;
}
.params::-webkit-scrollbar {
    width: 8px;
}
.params::-webkit-scrollbar-track {
    background: #2F2F37;
}
.params::-webkit-scrollbar-thumb {
    background-color: #8585AD;
}
.param {
    display: flex;
    flex-direction: row;
    flex-wrap: nowrap;
    justify-content: space-between;
    align-items: center;
    align-content: center;
}
.param:last-child {
    margin-bottom: 0px !important;
}
.title {
    display: flex;
    flex-direction: row;
    flex-wrap: wrap;
    justify-content: flex-start;
    align-items: center;
    align-content: center;
}
.icon {
    width: 64px;
    height: 64px;
    border-radius: 50%;
    border: solid 2px #7622BF;
    display: flex;
    flex-direction: row;
    flex-wrap: wrap;
    justify-content: center;
    align-items: center;
    align-content: center;
}
.icon img {
    height: 100%;
}
.bottom {
    display: flex;
    flex-direction: row;
    flex-wrap: wrap;
    justify-content: space-between;
    align-items: center;
    align-content: center;
}


