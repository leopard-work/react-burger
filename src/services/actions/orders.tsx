export const ORDERS_CONNECT: "ORDERS_CONNECT" = "ORDERS_CONNECT";
export const ORDERS_DISCONNECT: "ORDERS_DISCONNECT" = "ORDERS_DISCONNECT";
export const ORDERS_WS_CONNECTING: "ORDERS_WS_CONNECTING" =
  "ORDERS_WS_CONNECTING";
export const ORDERS_WS_OPEN: "ORDERS_WS_OPEN" = "ORDERS_WS_OPEN";
export const ORDERS_WS_CLOSE: "ORDERS_WS_CLOSE" = "ORDERS_WS_CLOSE";
export const ORDERS_WS_ERROR: "ORDERS_WS_ERROR" = "ORDERS_WS_ERROR";
export const ORDERS_WS_MESSAGE: "ORDERS_WS_MESSAGE" = "ORDERS_WS_MESSAGE";

export const OPEN_ORDER_ITEM: "OPEN_ORDER_ITEM" = "OPEN_ORDER_ITEM";
export const CLOSE_ORDER_ITEM: "CLOSE_ORDER_ITEM" = "CLOSE_ORDER_ITEM";

interface ORDERS_CONNECT_ACTION {
  readonly type: typeof ORDERS_CONNECT;
}
interface ORDERS_DISCONNECT_ACTION {
  readonly type: typeof ORDERS_DISCONNECT;
}
interface ORDERS_WS_CONNECTING_ACTION {
  readonly type: typeof ORDERS_WS_CONNECTING;
}
interface ORDERS_WS_OPEN_ACTION {
  readonly type: typeof ORDERS_WS_OPEN;
}
interface ORDERS_WS_CLOSE_ACTION {
  readonly type: typeof ORDERS_WS_CLOSE;
}
interface ORDERS_WS_ERROR_ACTION {
  readonly type: typeof ORDERS_WS_ERROR;
  payload: {
    error: string;
  };
}
interface ORDERS_WS_MESSAGE_ACTION {
  readonly type: typeof ORDERS_WS_MESSAGE;
  payload: {
    data: any;
  };
}
interface OPEN_ORDER_ITEM {
  readonly type: typeof OPEN_ORDER_ITEM;
}

interface CLOSE_ORDER_ITEM {
  readonly type: typeof CLOSE_ORDER_ITEM;
}

export type ORDERS_ACTIONS =
  | ORDERS_CONNECT_ACTION
  | ORDERS_DISCONNECT_ACTION
  | ORDERS_WS_CONNECTING_ACTION
  | ORDERS_WS_OPEN_ACTION
  | ORDERS_WS_CLOSE_ACTION
  | ORDERS_WS_ERROR_ACTION
  | ORDERS_WS_MESSAGE_ACTION
  | OPEN_ORDER_ITEM
  | CLOSE_ORDER_ITEM;
