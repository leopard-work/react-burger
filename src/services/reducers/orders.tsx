import {
  ORDERS_CONNECT,
  ORDERS_DISCONNECT,
  ORDERS_WS_CONNECTING,
  ORDERS_WS_OPEN,
  ORDERS_WS_CLOSE,
  ORDERS_WS_ERROR,
  ORDERS_WS_MESSAGE,
  ORDERS_ACTIONS,
  OPEN_ORDER_ITEM,
  CLOSE_ORDER_ITEM,
} from "../actions/orders";
import { OrderItemProps } from "../../utils/types";

type ordersState = {
  ordersData: {
    success: boolean;
    orders: any;
    total: number;
    totalToday: number;
  };
  status: string;
  connect: boolean;
  error: string;
  viewFullOrder: OrderItemProps;
  viewFullOrderModalOpen: boolean;
};

export const initialState: ordersState = {
  ordersData: {
    success: false,
    orders: [],
    total: 0,
    totalToday: 0,
  },
  status: "",
  connect: false,
  error: "",
  viewFullOrder: {
    _id: "",
    ingredients: [],
    status: "",
    name: "",
    createdAt: "",
    updatedAt: "",
    number: 0,
  },
  viewFullOrderModalOpen: false,
};

export const ordersReducer = (state = initialState, action: ORDERS_ACTIONS) => {
  switch (action.type) {
    case ORDERS_CONNECT: {
      return {
        ...state,
        status: "connect...",
      };
    }
    case ORDERS_DISCONNECT: {
      return initialState;
    }
    case ORDERS_WS_CONNECTING: {
      return {
        ...state,
        status: "connecting",
        connect: true,
      };
    }
    case ORDERS_WS_OPEN: {
      return {
        ...state,
        status: "open",
      };
    }
    case ORDERS_WS_MESSAGE: {
      return {
        ...state,
        status: "message",
        ordersData: action.data,
      };
    }
    case ORDERS_WS_CLOSE: {
      return {
        ...state,
        status: "close",
        connect: false,
      };
    }
    case ORDERS_WS_ERROR: {
      return {
        ...state,
        status: "error",
        error: action.error,
        connect: false,
      };
    }
    case OPEN_ORDER_ITEM: {
      return {
        ...state,
        viewFullOrderModalOpen: true,
        viewFullOrder: action.item,
      };
    }
    case CLOSE_ORDER_ITEM: {
      return {
        ...state,
        viewFullOrderModalOpen: false,
      };
    }
    default: {
      return state;
    }
  }
};
