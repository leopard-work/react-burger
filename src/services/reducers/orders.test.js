import {
  ORDERS_CONNECT,
  ORDERS_DISCONNECT,
  ORDERS_WS_CONNECTING,
  ORDERS_WS_OPEN,
  ORDERS_WS_CLOSE,
  ORDERS_WS_ERROR,
  ORDERS_WS_MESSAGE,
  ORDERS_ACTIONS,
  OPEN_ORDER_ITEM,
  CLOSE_ORDER_ITEM,
} from "../actions/orders";

import { initialState, ordersReducer } from "./orders";
import renderer from "react-test-renderer";
import React from "react";

const testOrder = {
  ingredients: [],
  _id: "",
  owner: {
    name: "",
    email: "",
    createdAt: "",
    updatedAt: "",
  },
  status: "",
  name: "",
  createdAt: "",
  updatedAt: "",
  number: null,
  price: 0,
};

describe("Лента заказов", () => {
  it("- исходное состояние", () => {
    expect(ordersReducer(initialState, {})).toEqual(initialState);
    const tree = renderer.create(<>Исходное состояние</>).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it("- обработка ORDERS_CONNECT", () => {
    expect(
      ordersReducer(initialState, {
        type: ORDERS_CONNECT,
      })
    ).toEqual({
      ...initialState,
      status: "connect...",
    });
    const tree = renderer.create(<>Подключение websocket</>).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it("- обработка ORDERS_WS_CONNECTING", () => {
    expect(
      ordersReducer(initialState, {
        type: ORDERS_WS_CONNECTING,
      })
    ).toEqual({
      ...initialState,
      status: "connecting",
      connect: true,
    });
    const tree = renderer.create(<>websocket подключен</>).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it("- обработка ORDERS_WS_OPEN", () => {
    expect(
      ordersReducer(initialState, {
        type: ORDERS_WS_OPEN,
      })
    ).toEqual({
      ...initialState,
      status: "open",
    });
    const tree = renderer.create(<>websocket открытие</>).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it("- обработка ORDERS_WS_MESSAGE", () => {
    expect(
      ordersReducer(initialState, {
        type: ORDERS_WS_MESSAGE,
      })
    ).toEqual({
      ...initialState,
      status: "message",
      ordersData: undefined,
    });
    const tree = renderer.create(<>websocket получение сообщения</>).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it("- обработка ORDERS_WS_CLOSE", () => {
    expect(
      ordersReducer(initialState, {
        type: ORDERS_WS_CLOSE,
      })
    ).toEqual({
      ...initialState,
      status: "close",
      connect: false,
    });
    const tree = renderer.create(<>websocket закрытие</>).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it("- обработка ORDERS_WS_ERROR", () => {
    expect(
      ordersReducer(initialState, {
        type: ORDERS_WS_ERROR,
        error: "test",
      })
    ).toEqual({
      ...initialState,
      status: "error",
      error: "test",
      connect: false,
    });
    const tree = renderer.create(<>websocket ошибка</>).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it("- обработка OPEN_ORDER_ITEM", () => {
    expect(
      ordersReducer(initialState, {
        type: OPEN_ORDER_ITEM,
        item: testOrder,
      })
    ).toEqual({
      ...initialState,
      viewFullOrder: testOrder,
      viewFullOrderModalOpen: true,
    });
    const tree = renderer.create(<>открытие отдельного заказа</>).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it("- обработка CLOSE_ORDER_ITEM", () => {
    expect(
      ordersReducer(initialState, {
        type: CLOSE_ORDER_ITEM,
      })
    ).toEqual({
      ...initialState,
      viewFullOrderModalOpen: false,
    });
    const tree = renderer.create(<>закрытие отдельного заказа</>).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
