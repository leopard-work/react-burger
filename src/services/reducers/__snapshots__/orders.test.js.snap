// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Лента заказов - исходное состояние 1`] = `"Исходное состояние"`;

exports[`Лента заказов - обработка CLOSE_ORDER_ITEM 1`] = `"закрытие отдельного заказа"`;

exports[`Лента заказов - обработка OPEN_ORDER_ITEM 1`] = `"открытие отдельного заказа"`;

exports[`Лента заказов - обработка ORDERS_CONNECT 1`] = `"Подключение websocket"`;

exports[`Лента заказов - обработка ORDERS_WS_CLOSE 1`] = `"websocket закрытие"`;

exports[`Лента заказов - обработка ORDERS_WS_CONNECTING 1`] = `"websocket подключен"`;

exports[`Лента заказов - обработка ORDERS_WS_ERROR 1`] = `"websocket ошибка"`;

exports[`Лента заказов - обработка ORDERS_WS_MESSAGE 1`] = `"websocket получение сообщения"`;

exports[`Лента заказов - обработка ORDERS_WS_OPEN 1`] = `"websocket открытие"`;
